File: hostel_management_system_backend/Services/AmenitiesService.cs
using AutoMapper;
using hostel_management_system_backend.Exceptions;
using Microsoft.EntityFrameworkCore;
using System.Globalization;

public interface IAmenitiesService
{
    Task<List<AmenityReadDto>> GetAllAsync(CancellationToken cancellationToken);
    Task<AmenityReadDto?> GetByIdAsync(Guid id, CancellationToken cancellationToken);
    Task<AmenityReadDto> CreateAsync(AmenityCreateDto dto, CancellationToken cancellationToken);
    Task<AmenityReadDto?> UpdateAsync(Guid id, AmenityUpdateDto dto, CancellationToken cancellationToken);
    Task<bool> DeleteAsync(Guid id, CancellationToken cancellationToken);
}

public sealed class AmenitiesService : IAmenitiesService
{
    private readonly ICrudRepository<Amenity> _repo;
    private readonly IMapper _mapper;

    public AmenitiesService(ICrudRepository<Amenity> repo, IMapper mapper)
    {
        _repo = repo;
        _mapper = mapper;
    }

    public async Task<List<AmenityReadDto>> GetAllAsync(CancellationToken cancellationToken)
    {
        var amenities = await _repo.GetAllAsNoTrackingAsync(cancellationToken);
        return _mapper.Map<List<AmenityReadDto>>(amenities.OrderBy(x => x.Name));
    }

    public async Task<AmenityReadDto?> GetByIdAsync(Guid id, CancellationToken cancellationToken)
    {
        var amenity = await _repo.GetByIdAsNoTrackingAsync(id, cancellationToken);
        if (amenity is null)
            throw new NotFoundException("Amenity not found.");

        return _mapper.Map<AmenityReadDto>(amenity);
    }

    public async Task<AmenityReadDto> CreateAsync(AmenityCreateDto dto, CancellationToken cancellationToken)
    {
        var names = SplitAmenityNames(dto.Name);
        if (names.Count == 0)
            throw new BadRequestException("Amenity name is required.");

        var existingAmenities = await _repo.GetAllAsNoTrackingAsync(cancellationToken);
        var existingNames = existingAmenities
            .Select(x => NormalizeName(x.Name))
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

        Amenity? firstCreated = null;
        foreach (var name in names)
        {
            var normalized = NormalizeName(name);
            if (existingNames.Contains(normalized))
                continue;

            var entity = new Amenity
            {
                Name = name,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = null,
                IsDeleted = false
            };

            await _repo.AddAsync(entity, cancellationToken);
            firstCreated ??= entity;
            existingNames.Add(normalized);
        }

        if (firstCreated is null)
            throw new ConflictException("An amenity with the same name already exists.", "amenity_name_conflict");

        try
        {
            await _repo.SaveChangesAsync(cancellationToken);
        }
        catch (DbUpdateException ex)
        {
            throw new ConflictException("An amenity with the same name already exists.", "amenity_name_conflict", ex);
        }

        return _mapper.Map<AmenityReadDto>(firstCreated);
    }

    public async Task<AmenityReadDto?> UpdateAsync(Guid id, AmenityUpdateDto dto, CancellationToken cancellationToken)
    {
        var entity = await _repo.GetByIdForUpdateAsync(id, cancellationToken);
        if (entity is null)
            throw new NotFoundException("Amenity not found.");

        var names = SplitAmenityNames(dto.Name);
        if (names.Count == 0)
            throw new BadRequestException("Amenity name is required.");

        var existingAmenities = await _repo.GetAllAsNoTrackingAsync(cancellationToken);
        var currentNormalizedName = NormalizeName(entity.Name);

        var primaryName = names[0];
        var primaryNormalized = NormalizeName(primaryName);
        var primaryAlreadyExistsOnAnother = existingAmenities.Any(x =>
            x.Id != entity.Id &&
            string.Equals(NormalizeName(x.Name), primaryNormalized, StringComparison.OrdinalIgnoreCase));

        if (primaryAlreadyExistsOnAnother)
            throw new ConflictException("An amenity with the same name already exists.", "amenity_name_conflict");

        entity.Name = primaryName;
        entity.UpdatedAt = DateTime.UtcNow;

        var existingNames = existingAmenities
            .Where(x => x.Id != entity.Id)
            .Select(x => NormalizeName(x.Name))
            .ToHashSet(StringComparer.OrdinalIgnoreCase);
        existingNames.Add(primaryNormalized);
        existingNames.Add(currentNormalizedName);

        foreach (var extraName in names.Skip(1))
        {
            var normalized = NormalizeName(extraName);
            if (existingNames.Contains(normalized))
                continue;

            await _repo.AddAsync(new Amenity
            {
                Name = extraName,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = null,
                IsDeleted = false
            }, cancellationToken);

            existingNames.Add(normalized);
        }

        await _repo.SaveChangesAsync(cancellationToken);
        return _mapper.Map<AmenityReadDto>(entity);
    }

    public async Task<bool> DeleteAsync(Guid id, CancellationToken cancellationToken)
    {
        var entity = await _repo.GetByIdForUpdateAsync(id, cancellationToken);
        if (entity is null)
            throw new NotFoundException("Amenity not found.");

        entity.IsDeleted = true;
        entity.DeletedAt = DateTime.UtcNow;
        entity.UpdatedAt = DateTime.UtcNow;
        await _repo.SaveChangesAsync(cancellationToken);
        return true;
    }

    private static List<string> SplitAmenityNames(string rawName)
    {
        if (string.IsNullOrWhiteSpace(rawName))
            return new List<string>();

        return rawName
            .Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries)
            .Select(x => x.Trim())
            .Where(x => !string.IsNullOrWhiteSpace(x))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();
    }

    private static string NormalizeName(string value)
    {
        return value.Trim().ToLower(CultureInfo.InvariantCulture);
    }
}

File: src/amenities/amenities.service.ts
import { Injectable } from '@nestjs/common';

import {
  AppBadRequestException,
  AppConflictException,
  AppNotFoundException,
} from '../common/exceptions/app-exception';
import { DatabaseService } from '../database/database.service';
import { AmenityCreateDto, AmenityReadDto, AmenityUpdateDto } from './dto/amenities.dto';

@Injectable()
export class AmenitiesService {
  constructor(private readonly db: DatabaseService) {}

  async getAll(): Promise<AmenityReadDto[]> {
    const amenities = await this.db.amenity.findMany({
      where: { isDeleted: false },
      orderBy: { name: 'asc' },
    });

    return amenities.map((amenity) => ({
      id: amenity.id,
      name: amenity.name,
      createdAt: amenity.createdAt,
      updatedAt: amenity.updatedAt,
    }));
  }

  async getById(id: string): Promise<AmenityReadDto> {
    const amenity = await this.db.amenity.findFirst({
      where: {
        id,
        isDeleted: false,
      },
    });

    if (!amenity) {
      throw new AppNotFoundException('Amenity not found.');
    }

    return {
      id: amenity.id,
      name: amenity.name,
      createdAt: amenity.createdAt,
      updatedAt: amenity.updatedAt,
    };
  }

  async create(dto: AmenityCreateDto): Promise<AmenityReadDto> {
    const names = this.splitAmenityNames(dto.name);
    if (!names.length) {
      throw new AppBadRequestException('Amenity name is required.');
    }

    const existingAmenities = await this.db.amenity.findMany({
      where: { isDeleted: false },
      select: { name: true },
    });
    const existingNames = new Set(existingAmenities.map((amenity) => this.normalizeName(amenity.name)));

    let firstCreated: AmenityReadDto | null = null;

    for (const name of names) {
      const normalized = this.normalizeName(name);
      if (existingNames.has(normalized)) {
        continue;
      }

      const amenity = await this.db.amenity.create({
        data: {
          name,
          createdAt: new Date(),
          isDeleted: false,
        },
      });

      existingNames.add(normalized);

      if (!firstCreated) {
        firstCreated = {
          id: amenity.id,
          name: amenity.name,
          createdAt: amenity.createdAt,
          updatedAt: amenity.updatedAt,
        };
      }
    }

    if (!firstCreated) {
      throw new AppConflictException('An amenity with the same name already exists.', 'amenity_name_conflict');
    }

    return firstCreated;
  }

  async update(id: string, dto: AmenityUpdateDto): Promise<AmenityReadDto> {
    const amenity = await this.db.amenity.findFirst({
      where: {
        id,
        isDeleted: false,
      },
    });

    if (!amenity) {
      throw new AppNotFoundException('Amenity not found.');
    }

    const names = this.splitAmenityNames(dto.name);
    if (!names.length) {
      throw new AppBadRequestException('Amenity name is required.');
    }

    const primaryName = names[0];
    const duplicate = await this.db.amenity.findFirst({
      where: {
        id: { not: id },
        isDeleted: false,
        name: {
          equals: primaryName,
          mode: 'insensitive',
        },
      },
    });

    if (duplicate) {
      throw new AppConflictException('An amenity with the same name already exists.', 'amenity_name_conflict');
    }

    const updated = await this.db.amenity.update({
      where: { id },
      data: {
        name: primaryName,
        updatedAt: new Date(),
      },
    });

    const existingAmenities = await this.db.amenity.findMany({
      where: {
        id: { not: id },
        isDeleted: false,
      },
      select: { name: true },
    });
    const existingNames = new Set(existingAmenities.map((item) => this.normalizeName(item.name)));
    existingNames.add(this.normalizeName(primaryName));
    existingNames.add(this.normalizeName(amenity.name));

    for (const extraName of names.slice(1)) {
      const normalized = this.normalizeName(extraName);
      if (existingNames.has(normalized)) {
        continue;
      }

      await this.db.amenity.create({
        data: {
          name: extraName,
          createdAt: new Date(),
          isDeleted: false,
        },
      });

      existingNames.add(normalized);
    }

    return {
      id: updated.id,
      name: updated.name,
      createdAt: updated.createdAt,
      updatedAt: updated.updatedAt,
    };
  }

  async delete(id: string) {
    const amenity = await this.db.amenity.findFirst({
      where: {
        id,
        isDeleted: false,
      },
    });

    if (!amenity) {
      throw new AppNotFoundException('Amenity not found.');
    }

    await this.db.amenity.update({
      where: { id },
      data: {
        isDeleted: true,
        deletedAt: new Date(),
        updatedAt: new Date(),
      },
    });
  }

  private splitAmenityNames(raw: string): string[] {
    return [...new Set((raw ?? '')
      .split(',')
      .map((value) => value.trim())
      .filter((value) => Boolean(value)))];
  }

  private normalizeName(value: string): string {
    return value.trim().toLowerCase();
  }
}
